multicluster/controllers/multicluster/stale_controller.go

```diff
@@ -102,128 +122,169 @@ func (c *StaleController) cleanupStaleServiceResources(remoteCluster commonarea.
-	for k, re := range resExpItems {
+	for i, resExp := range allResExpItems {
+		if (resExp.Spec.Kind == common.ServiceKind || resExp.Spec.Kind == common.EndpointsKind) && resExp.Labels[common.SourceClusterID] == c.localClusterID {
+			staleResExpItems = append(staleResExpItems, allResExpItems[i])
+		}
+	}
+
+	for k, re := range staleResExpItems {
 		for _, se := range svcExpItems {
 			if re.Spec.Name == se.Name && re.Spec.Namespace == se.Namespace {
 				// Set the valid ResourceExport item as empty ResourceExport, then all left non-empty items should be removed.
-				resExpItems[k] = mcsv1alpha1.ResourceExport{}
+				staleResExpItems[k] = mcsv1alpha1.ResourceExport{}
 			}
 		}
 	}
 
-	for _, r := range resExpItems {
+	for _, r := range staleResExpItems {
 		re := r
 		if re.Name != "" {
 			klog.InfoS("Cleaning up ResourceExport", "ResourceExport", klog.KObj(&re))
-			if err := remoteCluster.Delete(ctx, &re, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			if err := commonArea.Delete(ctx, &re, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
 				return err
 			}
 		}
 	}
 	return nil
 }
 
-func (c *StaleController) cleanupACNPResources(resImpList *mcsv1alpha1.ResourceImportList) error {
-	acnpList := &crdv1alpha1.ClusterNetworkPolicyList{}
-	if err := c.List(ctx, acnpList, &client.ListOptions{}); err != nil {
+// cleanupClusterInfoResourceExport removes any Gateway kind of ResourceExports when there is no
+// Gateway in the local cluster.
+func (c *StaleResCleanupController) cleanupClusterInfoResourceExport(commonArea commonarea.RemoteCommonArea,
+	resExpList *mcsv1alpha1.ResourceExportList) error {
+	var gws mcsv1alpha1.GatewayList
+	if err := c.Client.List(ctx, &gws, &client.ListOptions{}); err != nil {
 		return err
 	}
-	staleMCACNPItems := map[string]crdv1alpha1.ClusterNetworkPolicy{}
-	for _, acnp := range acnpList.Items {
-		if _, ok := acnp.Annotations[common.AntreaMCACNPAnnotation]; ok {
-			staleMCACNPItems[acnp.Name] = acnp
+
+	if len(gws.Items) == 0 {
+		ciExport := &mcsv1alpha1.ResourceExport{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: commonArea.GetNamespace(),
+				Name:      c.localClusterID + "-clusterinfo",
+			},
 		}
-	}
-	for _, resImp := range resImpList.Items {
-		if resImp.Spec.Kind == common.AntreaClusterNetworkPolicyKind {
-			acnpNameFromResImp := common.AntreaMCSPrefix + resImp.Spec.Name
-			if _, ok := staleMCACNPItems[acnpNameFromResImp]; ok {
-				delete(staleMCACNPItems, acnpNameFromResImp)
-			}
-		}
-	}
-	for _, stalePolicy := range staleMCACNPItems {
-		acnp := stalePolicy
-		klog.InfoS("Cleaning up stale ACNP", "acnp", klog.KObj(&acnp))
-		if err := c.Client.Delete(ctx, &acnp, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
-			return err
+		if err := commonArea.Delete(ctx, ciExport, &client.DeleteOptions{}); err != nil {
+			klog.V(2).InfoS("Cleaning up stale ClusterInfo kind of ResourceExport", "resourceexport", klog.KObj(ciExport))
```

Comment:
Why it doesn't log error and return error?

Reply:
Fixed